File: scala-stuff/src/main/scala/io/github/flyingcowmoomoo/arrowcatsfpguild/scala/Effects.scala
package io.github.flyingcowmoomoo.arrowcatsfpguild.scala

import cats.{Applicative, Monad}
import cats.effect.{Async, IO, IOApp}
import cats.effect.unsafe.implicits.global
import cats.implicits._

import java.util.concurrent.atomic.AtomicLong
import scala.collection.concurrent.TrieMap
import scala.concurrent.ExecutionContext
import scala.math.BigDecimal


object Effects{

  def main(args: Array[String]): Unit = {
    exec()
  }
  def exec(): Unit = {

    val productRepoIO: RepositoryAlgebra[IO] = new MyRepository[IO](new AtomicLong())

    val programIO = new MyProgram(productRepoIO)



    val finalProgram: IO[(Option[Product], Option[Product], Option[Product])] = for {
      createInvalidPrice <- programIO.createProduct("Test 1", 666)
      createValidPrice <- programIO.createProduct("Test 2", 1337)
      // Dirty below, can throw a NoSuchElementException
      getCreateValidPrice <- programIO.getProduct(createValidPrice.get.id.get)
    } yield (createInvalidPrice, createValidPrice, getCreateValidPrice)

    implicit val context: IO[ExecutionContext] =  Async[IO].executionContext

//    val parallelProgram = (
//      programIO.createProduct("Test 1", 666),
//      programIO.createProduct("Test 2", 1337)
//      ).parMapN((a, b) => (a, b, programIO.getProduct(b.get.id.get)))

    val syncResult = finalProgram.unsafeRunSync()
//    val parallelResult = parallelProgram.unsafeRunSync()
    val parallelResult = None

    val result = Map("syncResult" -> syncResult, "parallelResult" -> parallelResult)

    println(result)
  }

  class MyProgram[F[_]: Monad](repo: RepositoryAlgebra[F]) {
    def renameProduct(id: Long, title: String): F[Option[Product]] = repo
      .get(id)
      .flatMap {
        case Some(product) =>
          val renamed = product.copy(title = title)
          repo.update(renamed) map (_ => Option(renamed))
        case None =>
          Monad[F].pure(None)
      }

    def getProduct(id: Long): F[Option[Product]] = repo.get(id)

    def createProduct(title: String, price: BigDecimal): F[Option[Product]] = {
      if (price.equals(BigDecimal(666))) {
        Monad[F].pure(None)
      }
      else {
        repo.create(Product(None, title, price)) map (it => Option(it))
      }
    }

    def deleteProduct(id: Long): F[Option[Product]] = repo.delete(id)
  }
}

case class Product(id: Option[Long], title: String, price: BigDecimal)

trait RepositoryAlgebra[F[_]] {
  def create(product: Product): F[Product]

  def update(product: Product): F[Option[Product]]

  def get(id: Long): F[Option[Product]]

  def delete(id: Long): F[Option[Product]]
}

class MyRepository[F[_] : Applicative](sequence: AtomicLong) extends RepositoryAlgebra[F] {
  private val cache = TrieMap[Long, Product]()

  def create(product: Product): F[Product] = {
    val id: Long = sequence.getAndIncrement()
    val toSave = product.copy(id = id.some)
    cache += (id -> toSave)
    toSave.pure[F]
  }

  override def update(product: Product): F[Option[Product]] = product.id.traverse { id =>
    cache.update(id, product)
    product.pure[F]
  }

  override def get(id: Long): F[Option[Product]] = cache.get(id).pure[F]

  override def delete(id: Long): F[Option[Product]] = cache.remove(id).pure[F]
}

File: kotlin-stuff/src/main/kotlin/io/github/flyingcowmoomoo/kotlin/arrowcatsfpguild/Effects.kt
package io.github.flyingcowmoomoo.kotlin.arrowcatsfpguild

import arrow.core.Option
import arrow.core.none
import arrow.core.some
import arrow.core.toOption
import arrow.fx.coroutines.parZip
import kotlinx.coroutines.runBlocking
import java.math.BigDecimal
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.atomic.AtomicLong


fun main() {
    val repo = MyRepository(AtomicLong())

    val program = MyProgram(repo)


     val finalProgram: suspend () -> Triple<Option<Product>, Option<Product>, Option<Product>> = suspend {
         val createInvalidPrice: Option<Product> = program.createProduct("Test 1", BigDecimal(666))
         val createValidPrice = program.createProduct("Test 2", BigDecimal(1337))
         val getCreateValidPrice = program.getProduct(createValidPrice.fold({-1}, {it.id.fold({-1}, { that -> that})}))
         Triple(createInvalidPrice,createValidPrice, getCreateValidPrice)
     }

    val parallelProgram: suspend () -> Triple<Option<Product>, Option<Product>, Option<Product>> = suspend {
        parZip(
            { program.createProduct("Test 1", BigDecimal(666)) },
            { program.createProduct("Test 2", BigDecimal(1337))  },
        ) { a, b ->
            Triple(a, b, program.getProduct(b.fold({-1}, {it.id.fold({-1}, { that -> that})})))
        }
    }

    val result = runBlocking {
        runBlocking {
            val syncResult = finalProgram.invoke()
            val parallelResult = parallelProgram.invoke()
            mapOf(Pair("syncResult", syncResult), Pair("parallelResult", parallelResult))
        }
    }

     println(result)
}

object Effects {

}
data class Product(val id: Option<Long>, val title: String, val price: BigDecimal)


class MyProgram(private val repo: Repository) {
    suspend fun renameProduct(id: Long, title: String): Option<Product> {
        return repo
            .get(id)
            .flatMap {
                val renamed = it.copy(title = title)
                repo.update(renamed)
            }
    }

    fun getProduct(id: Long): Option<Product> = repo.get(id)

    suspend fun createProduct(title: String, price: BigDecimal): Option<Product> {
        return when (price) {
            BigDecimal(666) -> none()
            else -> repo.create(Product(none(), title, price)).toOption()
        }
    }

    suspend fun deleteProduct(id: Long): Option<Product> = repo.delete(id)
}

class MyRepository(private val sequence: AtomicLong): Repository {

    private val cache = ConcurrentHashMap<Long, Product>()

    override fun create(product: Product): Product {
        val id = sequence.getAndIncrement()
        val toSave = product.copy(id = id.some())
        cache[id] = toSave
        return toSave
    }

    override fun update(product: Product): Option<Product> {
        return product.id.fold({ none() }, { cache.replace(it, product).toOption() })
    }

    override fun get(id: Long): Option<Product> {
        return cache[id].toOption()
    }

    override fun delete(id: Long): Option<Product> {
        return cache.remove(id).toOption()
    }

}
interface Repository {
    fun create(product: Product): Product
    fun update(product: Product): Option<Product>
    fun get(id: Long): Option<Product>
    fun delete(id: Long): Option<Product>
}